### PR TITLE
WLTS- 5 : Build new GET /dictionary endpoint

### PR DESCRIPTION
**Pull request by sullivangareth**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/WLTS-5

### Change description ###

download a set of phrases associated with potentially associated with a case type for which translations will be eventually added.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X] No
```

### CODE REVIEW

**Review by danlysiak — COMMENTED** [review]
Couple of little suggestions

File: src/main/java/uk/gov/hmcts/reform/translate/model/GetDictionaryResponse.java [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+import lombok.AllArgsConstructor;
+import lombok.Data;
+
+import java.util.Map;
+
+@AllArgsConstructor
+@Data
+public class GetDictionaryResponse {
+    private Map<String, String> translations;
+}
```
Comment (Thor-tech-of-metal) [inline src/main/java/uk/gov/hmcts/reform/translate/model/GetDictionaryResponse.java:12]: According to the spec It seems that PUT service and GET are sharing the same payload for request / response .  Could we rename GetDictionaryResponse?   or create an superclass with the main Map  attribute so then we can have concrete implementations for Get-response and Put-request ?

File: src/main/java/uk/gov/hmcts/reform/translate/controllers/DictionaryController.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+    private final DictionaryService dictionaryService;
+
+    @Autowired
+    public DictionaryController(DictionaryService dictionaryService) {
+        this.dictionaryService = dictionaryService;
+    }
+
+    @GetMapping(path = "/dictionary", produces = MediaType.APPLICATION_JSON_VALUE)
+    @ResponseStatus(HttpStatus.OK)
```
Comment (Thor-tech-of-metal) [inline src/main/java/uk/gov/hmcts/reform/translate/controllers/DictionaryController.java:26]: Apparently GET / PUT and other services are going to use as unique resource /dictionary. Then the hppt verbs GET PUT PATH etc are going to identify the  restful resource. Could we move /dictionary to the to of the controller in order to following the design and avoid duplications .

File: src/main/java/uk/gov/hmcts/reform/translate/controllers/DictionaryController.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    @ApiResponses(value = {
+        @ApiResponse(responseCode = "200", description = "Dictionary returned successfully"),
+        @ApiResponse(responseCode = "401", description = "Unauthorised"),
+        @ApiResponse(responseCode = "403", description = "Forbidden")
+    })
+    public ResponseEntity<GetDictionaryResponse> getDictionary() {
```
[the remark this answers is not in the document]
Comment (sullivangareth) [reply src/main/java/uk/gov/hmcts/reform/translate/controllers/DictionaryController.java:32]: ok, will fix

File: src/main/java/uk/gov/hmcts/reform/translate/service/DictionaryService.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        final var dictionaryEntities = dictionaryRepository.findAll();
+
+        final var spliterator = dictionaryEntities.spliterator();
+
+        Map<String, String> returnValue = Collections.emptyMap();
+
+        if (dictionaryEntities.spliterator() != null) {
+            Stream<DictionaryEntity> stream = StreamSupport.stream(spliterator, false);
+
+            returnValue = stream.collect(Collectors.toMap(
```
[the remark this answers is not in the document]
Comment (sullivangareth) [reply src/main/java/uk/gov/hmcts/reform/translate/service/DictionaryService.java:34]: This is what I had originally - but I have a pet hate of multiple return statements in a method.  I will fix though !

File: src/test/java/uk/gov/hmcts/reform/translate/service/DictionaryServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    DictionaryRepository dictionaryRepository;
+
+    @Mock
+    Iterable<DictionaryEntity> repositoryResults;
+
+    @InjectMocks
+    DictionaryService dictionaryService;
+
+    @Test
+    void shouldReturnDictionaryContents() {
```
[the remark this answers is not in the document]
Comment (sullivangareth) [reply src/test/java/uk/gov/hmcts/reform/translate/service/DictionaryServiceTest.java:33]: OK, will add - I think the integration test deals with translationPhrase is null - but you are correct, it is also required here

Will also add duplicate records tests here and in IT

File: src/main/java/uk/gov/hmcts/reform/translate/service/DictionaryService.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+                    DictionaryEntity::getEnglishPhrase,
+                    dictionaryEntity ->
+                        dictionaryEntity.getTranslationPhrase() == null ? "" : dictionaryEntity.getTranslationPhrase()
```
[the remark this answers is not in the document]
Comment (sullivangareth) [reply src/main/java/uk/gov/hmcts/reform/translate/service/DictionaryService.java:37]: This is to workaround a known bug in Java - Collectors.toMap() throws NPE if value is null

https://bugs.openjdk.java.net/browse/JDK-8148463

File: src/main/java/uk/gov/hmcts/reform/translate/controllers/DictionaryController.java (hunk 2 above)
Comment (sullivangareth) [reply src/main/java/uk/gov/hmcts/reform/translate/controllers/DictionaryController.java:26]: will fix

File: src/main/java/uk/gov/hmcts/reform/translate/model/GetDictionaryResponse.java (hunk 1 above)
Comment (sullivangareth) [reply src/main/java/uk/gov/hmcts/reform/translate/model/GetDictionaryResponse.java:12]: Will rename to `DictionaryResponse`

File: src/main/java/uk/gov/hmcts/reform/translate/controllers/DictionaryController.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (sullivangareth) [reply src/main/java/uk/gov/hmcts/reform/translate/controllers/DictionaryController.java]: Fixed in https://github.com/hmcts/ts-translation-service/pull/6/commits/46585f3e1763b52bd2709a72f4a1cfc2f0ebc77b

File: src/main/java/uk/gov/hmcts/reform/translate/service/DictionaryService.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (sullivangareth) [reply src/main/java/uk/gov/hmcts/reform/translate/service/DictionaryService.java]: Fixed in https://github.com/hmcts/ts-translation-service/pull/6/commits/46585f3e1763b52bd2709a72f4a1cfc2f0ebc77b

File: src/test/java/uk/gov/hmcts/reform/translate/service/DictionaryServiceTest.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (sullivangareth) [reply src/test/java/uk/gov/hmcts/reform/translate/service/DictionaryServiceTest.java:36]: Unit tests added in  https://github.com/hmcts/ts-translation-service/pull/6/commits/46585f3e1763b52bd2709a72f4a1cfc2f0ebc77b

Unsure about the integration tests, if the `IllegalStateException` is thrown, what status code gets returned in the response?  

Would the  `PUT` endpoint perform this duplicate check and error, or would it overwrite the existing entry rather than add a duplicate ?

File: src/main/java/uk/gov/hmcts/reform/translate/model/GetDictionaryResponse.java (hunk 1 above)
Comment (sullivangareth) [reply src/main/java/uk/gov/hmcts/reform/translate/model/GetDictionaryResponse.java]: Fixed in https://github.com/hmcts/ts-translation-service/pull/6/commits/46585f3e1763b52bd2709a72f4a1cfc2f0ebc77b

File: src/test/java/uk/gov/hmcts/reform/translate/service/DictionaryServiceTest.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (danlysiak) [reply src/test/java/uk/gov/hmcts/reform/translate/service/DictionaryServiceTest.java:36]: I'd expect the `IllegalStateException` is likely returning a `500`, as we're not converting it anywhere. Which is OK - if we somehow found ourselves in that situation then there's nothing the consumer can do to rectify, it'd require manual intervention.
The `PUT` logic is described such that it would overwrite.

File: src/main/java/uk/gov/hmcts/reform/translate/data/DictionaryEntity.java [hunk 7]
```diff
@@ -15,6 +17,8 @@
 
 @Table(name = "dictionary")
 @Entity
+@Builder
+@AllArgsConstructor
 @NoArgsConstructor
 @Data
```
Comment (Thor-tech-of-metal) [inline src/main/java/uk/gov/hmcts/reform/translate/data/DictionaryEntity.java:23]: We are using 2 lombok annotations to add generated code for all propouses for instance : @AllArgsConstructor
@NoArgsConstructor they are the opposite . May be @Data is good enough

Comment (Thor-tech-of-metal) [reply src/main/java/uk/gov/hmcts/reform/translate/data/DictionaryEntity.java:23]: I will do  these changes on my branch.

**Review by Thor-tech-of-metal — APPROVED** [review]
LGTM

File: charts/ts-translation-service/values.preview.template.yaml [hunk 8]
```diff
@@ -2,3 +2,17 @@ java:
...
+    TS_TRANSLATION_SERVICE_DB_PASSWORD: "{{ .Values.postgresql.postgresqlPassword}}"
+    TS_TRANSLATION_SERVICE_DB_OPTIONS: "?stringtype=unspecified"
+    TS_TRANSLATION_SERVICE_DB_PORT: 5432
+  postgresql:
+    enabled: true
+  keyVaults:
+    ts:
+      resourceGroup: ts-shared
+      secrets:
+        - name: AppInsightsInstrumentationKey
```
Comment (danlysiak) [inline charts/ts-translation-service/values.preview.template.yaml:18]: I suspect we need to also add `IDAM_KEY_TS_TRANSLATION_SERVICE` and `TS_TRANSLATION_SERVICE_IDAM_CLIENT_SECRET` for preview to work fully.

File: src/integrationTest/java/uk/gov/hmcts/reform/translate/controllers/DictionaryControllerIT.java [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+            mockMvc.perform(get(URL)
+                                .contentType(MediaType.APPLICATION_JSON_VALUE))
+                .andExpect(status().is(200))
+                .andExpect(jsonPath("$.translations", equalTo(expectedDictionary)))
+                .andReturn();
+        }
+
+        @Test
+        @Sql(scripts = {DELETE_TRANSLATION_TABLES_SCRIPT, GET_TRANSLATION_TABLES_DUPLICATE_ENGLISH_PHRASES_SCRIPT})
+        void shouldReturn500WhenDictionaryReturnsDuplicateEnglishPhrases() {
```
Comment (danlysiak) [inline src/integrationTest/java/uk/gov/hmcts/reform/translate/controllers/DictionaryControllerIT.java:65]: We should probably add assertions for the actual response and confirm the 500 in the test too.

File: src/main/java/uk/gov/hmcts/reform/translate/security/SecurityUtils.java [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+    public String getUserRolesHeader() {
+        Collection<? extends GrantedAuthority> authorities =
+            SecurityContextHolder.getContext().getAuthentication().getAuthorities();
+        return authorities.stream()
+            .map(GrantedAuthority::getAuthority)
+            .collect(Collectors.joining(","));
+    }
+
+    public boolean hasRole(String roleToMatch, List<String> userRoles) {
+        return userRoles.stream().anyMatch(role -> role.equalsIgnoreCase(roleToMatch));
```
Comment (danlysiak) [inline src/main/java/uk/gov/hmcts/reform/translate/security/SecurityUtils.java:82]:
```suggestion
        return getRoles().stream().anyMatch(role -> role.equalsIgnoreCase(roleToMatch));
```
Could make the second argument redundant, as it looks like we'll always be calling it in the context of the user in the context.

File: src/main/java/uk/gov/hmcts/reform/translate/security/SecurityUtils.java [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+
+    public String getUserRolesHeader() {
+        Collection<? extends GrantedAuthority> authorities =
+            SecurityContextHolder.getContext().getAuthentication().getAuthorities();
+        return authorities.stream()
+            .map(GrantedAuthority::getAuthority)
+            .collect(Collectors.joining(","));
+    }
+
+    public boolean hasRole(String roleToMatch, List<String> userRoles) {
```
Comment (danlysiak) [inline src/main/java/uk/gov/hmcts/reform/translate/security/SecurityUtils.java:81]: Could make the second argument redundant, as it looks like we'll always be calling it in the context of the user in the context?

File: src/main/java/uk/gov/hmcts/reform/translate/security/SecurityUtils.java (hunk 10 above)
Comment (danlysiak) [inline src/main/java/uk/gov/hmcts/reform/translate/security/SecurityUtils.java:82]: Could make the second argument redundant, as it looks like we'll always (in future too) be calling it for the user in the context?
```suggestion
    public boolean hasRole(String roleToMatch, List<String> userRoles) {
        return userRoles.stream().anyMatch(role -> role.equalsIgnoreCase(roleToMatch));
```

Comment (danlysiak) [inline src/main/java/uk/gov/hmcts/reform/translate/security/SecurityUtils.java:82]: Could make the second argument redundant, as it looks like we'll always (in future too) be calling it for the user in the context?
```suggestion
    public boolean hasRole(String roleToMatch) {
        UserInfo userInfo = getUserInfo();
        return userInfo != null && userInfo.getRoles().stream().anyMatch(role -> role.equalsIgnoreCase(roleToMatch));
```

File: src/cftlib/java/uk/gov/hmcts/reform/translate/CftLibConfig.java [hunk 12]
```diff
@@ -0,0 +1,29 @@
+package uk.gov.hmcts.reform.translate;
+
+import org.springframework.stereotype.Component;
+import uk.gov.hmcts.rse.ccd.lib.api.CFTLib;
+import uk.gov.hmcts.rse.ccd.lib.api.CFTLibConfigurer;
+
+@Component
+public class CftLibConfig implements CFTLibConfigurer {
```
[the remark this answers is not in the document]
Comment (sullivangareth) [reply src/cftlib/java/uk/gov/hmcts/reform/translate/CftLibConfig.java]: Committed in error, removed in https://github.com/hmcts/ts-translation-service/pull/6/commits/8a96c808b586488a19e17ec93b94553d5a8cd06e

File: src/integrationTest/java/uk/gov/hmcts/reform/translate/controllers/DictionaryControllerIT.java (hunk 9 above)
Comment (sullivangareth) [reply src/integrationTest/java/uk/gov/hmcts/reform/translate/controllers/DictionaryControllerIT.java:65]: Im not sure its actually throwing a 500? If I change this method to 

```
 mockMvc.perform(get(URL)
                                .contentType(MediaType.APPLICATION_JSON_VALUE))
                .andExpect(status().is(500))
                .andReturn();
```

it fails with `Request processing failed; nested exception is java.lang.IllegalStateException`

File: src/main/java/uk/gov/hmcts/reform/translate/security/SecurityUtils.java (hunk 10 above)
Comment (sullivangareth) [reply src/main/java/uk/gov/hmcts/reform/translate/security/SecurityUtils.java]: Fixed in [87b69a8](https://github.com/hmcts/ts-translation-service/pull/6/commits/87b69a874a6a8f9d6d484b2ad74838e9e425c207)

File: src/integrationTest/java/uk/gov/hmcts/reform/translate/controllers/DictionaryControllerIT.java [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.hamcrest.CoreMatchers.is;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+@AutoConfigureMockMvc(addFilters = false)
```
[the remark this answers is not in the document]
Comment (sullivangareth) [reply src/integrationTest/java/uk/gov/hmcts/reform/translate/controllers/DictionaryControllerIT.java]: Fixed in [87b69a8](https://github.com/hmcts/ts-translation-service/pull/6/commits/87b69a874a6a8f9d6d484b2ad74838e9e425c207)

File: src/integrationTest/java/uk/gov/hmcts/reform/translate/controllers/DictionaryControllerIT.java (hunk 9 above)
Comment (sullivangareth) [reply src/integrationTest/java/uk/gov/hmcts/reform/translate/controllers/DictionaryControllerIT.java]: method renamed to `shouldThrowExceptionWhenDictionaryReturnsDuplicateEnglishPhrases` in [87b69a8](https://github.com/hmcts/ts-translation-service/pull/6/commits/87b69a874a6a8f9d6d484b2ad74838e9e425c207)

File: charts/ts-translation-service/values.preview.template.yaml (hunk 8 above)
Comment (sullivangareth) [reply charts/ts-translation-service/values.preview.template.yaml:18]: Neither of these values exist yet in the `ts-aat` vault, so will add in a separate PR

**Review by danlysiak — APPROVED** [review]
